automerge-protocol/src/lib.rs

```diff
@@ -227,15 +244,134 @@ impl DataType {
+#[derive(PartialEq, Clone, Debug)]
+pub struct ScalarValues {
+    // For implementing Serialization in DiffEdit
+    pub(crate) vec: Vec<ScalarValue>,
+    // Can't use `std::mem::Discriminant` b/c we
+    // need to be able to `match` on the kind for `as_numerical_datatype`...
+    pub(crate) kind: ScalarValueKind,
+}
+
+impl ScalarValues {
+    pub fn from_values_and_datatype<'de, V: MapAccess<'de>>(
+        mut old_values: Vec<ScalarValue>,
+        datatype: Option<DataType>,
+    ) -> Result<Self, V::Error> {
+        // ensure the values can be cast to the correct datatype
+        if let Some(datatype) = datatype {
+            old_values = old_values
+                .iter()
+                .map(|v| {
+                    v.as_datatype(datatype).map_err(|e| {
+                        Error::invalid_value(
+                            Unexpected::Other(e.unexpected.as_str()),
+                            &e.expected.as_str(),
+                        )
+                    })
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+        }
+        old_values.try_into().map_err(|e| match e {
+            InvalidScalarValues::Empty => Error::invalid_length(0, &"more than 0"),
+            InvalidScalarValues::UnexpectedKind(exp, unexp) => {
+                let unexp = format!("{:?}", unexp);
+                let exp = format!("{:?}", exp);
+                Error::invalid_value(Unexpected::Other(&unexp), &exp.as_str())
+            }
+        })
+    }
+
+    /// Try to append a `ScalarValue` to a `ScalarValues`. If we can't
+    // returh the `ScalarValueKind` of the value we tried to add (for error reporting)
+    pub fn append(&mut self, v: ScalarValue) -> Option<ScalarValueKind> {
+        let new_kind = ScalarValueKind::from(&v);
+        if self.kind == new_kind {
+            self.vec.push(v);
+            None
+        } else {
+            Some(new_kind)
+        }
+    }
+
+    pub fn get(&self, idx: usize) -> Option<&ScalarValue> {
+        self.vec.get(idx)
+    }
+
+    pub fn len(&self) -> usize {
+        self.vec.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.vec.is_empty()
+    }
+
+    pub fn iter(&self) -> Iter<ScalarValue> {
+        self.vec.iter()
+    }
+
+    /// Returns an Option containing a `DataType` if
+    /// `self` represents a numerical scalar value
+    /// This is necessary b/c numerical values are not self-describing
+    /// (unlike strings / bytes / etc. )
+    pub fn as_numerical_datatype(&self) -> Option<DataType> {
+        match self.kind {
+            ScalarValueKind::Counter => Some(DataType::Counter),
+            ScalarValueKind::Timestamp => Some(DataType::Timestamp),
+            ScalarValueKind::Int => Some(DataType::Int),
+            ScalarValueKind::Uint => Some(DataType::Uint),
+            ScalarValueKind::F64 => Some(DataType::F64),
+            _ => None,
+        }
+    }
+}
+
+impl TryFrom<Vec<ScalarValue>> for ScalarValues {
```

Comment:
Can we move this and the `From` and `fmt::Display` impls below to `utility_impls`